### DISCUSSION TITLE
Formula fails to build but workflow succeeds

### DISCUSSION THREAD

**Discussion by dpo** [Homebrew/discussions#2929] [Tap maintenance and Homebrew development] [answered]
I setup CI  for the coin-or-tools/coinor tap using the yml files that Homebrew creates when initializing a new tap. Upon my initial PR where I iteratively fix the formulae to keep test-bot happy, I noticed that the workflow can be marked as a success even though certain formulae fail to build. Here is an example:

https://github.com/coin-or-tools/homebrew-coinor/runs/5015187989?check_suite_focus=true#step:8:292

In the Action summary, the build error only appears as a warning:

https://github.com/coin-or-tools/homebrew-coinor/actions/runs/1775923851

The `brew test-bot --only-formulae` job itself is marked as a success.

I find that that somewhat defeats the purpose of CI and requires manual examination of the logs, even in case of success. Is there an easy way to make the job and workflow fail when a formula fails to build?

Many thanks!

**Comment by carlocab** [comment, marked as answer]
The “warning” behaviour only kicks in when a formula has no bottle on the given OS. If it only warns on build failure when a formula has a pre-existing bottle, that’s a bug.

Still, it’s not the first I’ve heard of a request to change this behaviour. (Hi @ipatch.) A `-Werror`-type flag for `test-bot` should do this and be relatively easy to implement. (We probably won’t want to be selective about which warnings we promote to errors, though, as that would make things too complicated.) Happy to review a PR that does this.

**Reply by dpo** [reply]
Touché! :-) Many thanks for your help. I haven't been able to verify that there is indeed an error if a bottle already exists. If so, that will help greatly.

I'll keep the PR in mind but the Homebrew code is too sophisticated for my babbling Ruby. Do you have a pointer to get started?

**Reply by carlocab** [reply]
All decisions to ignore a failure flow through here: https://github.com/Homebrew/homebrew-test-bot/blob/e486c73e3067672689ab9d0c1c50fe2a2d7e6c62/lib/step.rb#L157-L163

It should be straightforward to add a command-line switch to override ignoring and just fail. It might take a little bit of work getting `test-bot` flags into the `Step` class, however.
